pydm/widgets/analog_indicator.py

```diff
@@ -0,0 +1,1576 @@
+
+        """
+        sets the pen color to alarm if the value is in the lower major alarm region
+        """
+        if self._value < self._lower_major_alarm:
+            self._painter.setBrush(self._major_alarm_color)
+        else:
+            self._painter.setBrush(self._major_alarm_region_color)
+
+
+        self._painter.drawRect(0, 15, lower_major_alarm_width, major_alarm_height)
+
+        """
+        sets the pen color to alarm if the value is in the upper major alarm region
+        """
+        if self._value > self._upper_major_alarm:
+            self._painter.setBrush(self._major_alarm_color)
+        else:
+            self._painter.setBrush(self._major_alarm_region_color)
+
+        self._painter.drawRect(upper_major_alarm_start, 15, upper_major_alarm_width, major_alarm_height)
+
+    def paintEvent(self, event):
+        """
+        Paint events are sent to widgets that need to update themselves,
+        for instance when part of a widget is exposed because a covering
+        widget was moved.
+
+        Parameters
+        ----------
+        event : QPaintEvent
+        """
+        self.adjust_transformation()
+        self._painter.begin(self)
+        self._painter.translate(0, self._painter_translation_y) # Draw vertically if needed
+        self._painter.rotate(self._painter_rotation)
+        self._painter.translate(self._painter_translation_x, 0) # Invert appearance if needed
+        self._painter.scale(self._painter_scale_x, 1)
+
+        self._painter.translate(0, self._flip_traslation_y)     # Invert scale if needed
+        self._painter.scale(1, self._flip_scale_y)
+
+        self._painter.setRenderHint(QPainter.Antialiasing)
+
+        """
+        self.draw_indicator() -> draws a triangle outside of the scale, used as the value indicator
+        self.draw_background() -> draws the background of the scale
+        self.draw_major() -> draws the major alarm area
+        self.draw_minor() -> draws the minor alarm area
+        self.draw_ticks() -> draws scale ticks
+        self.draw_set_point -> draws diamond in the middle of the scale. can use current self.draw_indicator() for this
+        """
+        self.draw_background()
+        self.draw_ticks()
+
+        self.draw_minor_alarm_region()
+        self.draw_major_alarm_region()
+        self.draw_indicator()
+        #self.draw_set_point()
+
+        self._painter.end()
+
+    def calculate_position_for_value(self, value):
+        """
+        Calculate the position (pixel) in which the pointer should be drawn for a given value.
+        """
+        if value is None or value < self._lower_limit or value > self._upper_limit or \
+           self._upper_limit - self._lower_limit == 0:
+            proportion = -1 # Invalid
+        else:
+            proportion = (value - self._lower_limit) / (self._upper_limit - self._lower_limit)
+
+        position = int(proportion * self._widget_width)
```

Comment:
I ran into this problem while taking it for a spin. I dragged the widget into an empty designer ui. I added a channel to an analog PV. The widget had no drawn bar so I went to make it bigger, and then this happened before it could render the bar:
```
Exception occurred while running Qt Designer.
Traceback (most recent call last):
  File "/cds/home/z/zlentz/pydev/pydm/widgets/analog_indicator.py", line 277, in paintEvent
    self.draw_minor_alarm_region()
  File "/cds/home/z/zlentz/pydev/pydm/widgets/analog_indicator.py", line 184, in draw_minor_alarm_region
    lower_minor_alarm_width = self.calculate_position_for_value(self._lower_minor_alarm)
  File "/cds/home/z/zlentz/pydev/pydm/widgets/analog_indicator.py", line 294, in calculate_position_for_value
    position = int(proportion * self._widget_width)
ValueError: cannot convert float NaN to integer
```

Reply:
This crashes the entire designer session very dramatically btw

Reply:
I switched to a different PV and it was fine- the PV that failed had the following hilarious metadata. I suggest finding a way to handle these in a non-crashing way:
```
$ caget -d 30 XPP:USR:MMS:01.RBV
XPP:USR:MMS:01.RBV
    Native data type: DBF_DOUBLE
    Request type:     DBR_CTRL_FLOAT
    Element count:    1
    Value:            -70.3204
    Status:           COMM
    Severity:         INVALID
    Units:            mm
    Precision:        4
    Lo disp limit:    -50
    Hi disp limit:    150
    Lo alarm limit:   nan
    Lo warn limit:    nan
    Hi warn limit:    nan
    Hi alarm limit:   nan
    Lo ctrl limit:    -50
    Hi ctrl limit:    150
```

Reply:
Hi Zach, thanks for all of your detailed feedback! 
I want to confirm that this crash is in designer and not a live PyDM display?  I want to try to replicate and fix this, however, I'm not sure what would cause a crash inside of designer as the alarm meta data should not be being pulled in in a designer session.
Thanks again!

Reply:
Yes, this crash occurs in designer. I haven't checked normal pydm yet, I was fixated on the editor crash.

> the alarm meta data should not be being pulled in in a designer session

Are you sure about this? We get the live values in designer to help the user see what their widget looks like using real data, so I assumed that all the metadata was expected too.

Reply:
Fixed!